src/rust/engine/fs/src/snapshot.rs

```diff
@@ -0,0 +1,352 @@
+    ).unwrap();
+    assert_eq!(
+      Snapshot::from_path_stats(store, digester, path_stats.clone())
+        .wait()
+        .unwrap(),
+      Snapshot {
+        fingerprint: fingerprint,
+        digest: Some(Digest(fingerprint, 78)),
+        path_stats: path_stats,
+      }
+    );
+  }
+
+  #[test]
+  fn snapshot_recursive_directories_including_empty() {
+    let (store, dir, posix_fs, digester) = setup();
+
+    let cats = PathBuf::from("cats");
+    let roland = cats.join("roland");
+    let dogs = PathBuf::from("dogs");
+    let llamas = PathBuf::from("llamas");
+    std::fs::create_dir_all(&dir.path().join(&cats)).unwrap();
+    std::fs::create_dir_all(&dir.path().join(&dogs)).unwrap();
+    std::fs::create_dir_all(&dir.path().join(&llamas)).unwrap();
+    make_file(&dir.path().join(&roland), STR.as_bytes(), 0o600);
+    let dogs_path_stat = PathStat::dir(
+      dogs.clone(),
+      downcast_dir_stat(posix_fs.stat(dogs).unwrap()),
+    );
+    let cats_roland_path_stat = PathStat::file(
+      roland.clone(),
+      downcast_file_stat(posix_fs.stat(roland).unwrap()),
+    );
+    let llamas_path_stat = PathStat::dir(
+      llamas.clone(),
+      downcast_dir_stat(posix_fs.stat(llamas).unwrap()),
+    );
+
+    let path_stats = vec![
+      dogs_path_stat.clone(),
+      cats_roland_path_stat.clone(),
+      llamas_path_stat.clone(),
+    ];
+    let sorted_path_stats = vec![cats_roland_path_stat, dogs_path_stat, llamas_path_stat];
+    // TODO: Inline when only used once
+    let fingerprint = Fingerprint::from_hex_string(
+      "fbff703bdaac62accf2ea5083bcfed89292073bf710ef9ad14d9298c637e777b",
+    ).unwrap();
+    assert_eq!(
+      Snapshot::from_path_stats(store, digester, path_stats.clone())
+        .wait()
+        .unwrap(),
+      Snapshot {
+        fingerprint: fingerprint,
+        digest: Some(Digest(fingerprint, 232)),
+        path_stats: sorted_path_stats,
+      }
+    );
+  }
+
+  struct FileSaver(Arc<Store>, Arc<PosixFS>);
+
+  impl GetFileDigest<String> for FileSaver {
+    fn digest(&self, file: &File) -> BoxFuture<Digest, String> {
+      let file_copy = file.clone();
+      let store = self.0.clone();
+      self
+        .1
+        .clone()
+        .read_file(&file)
+        .map_err(move |err| {
+          format!("Error reading file {:?}: {}", file_copy, err.description())
+        })
+        .and_then(move |content| store.store_file_bytes(&content.content))
+        .to_boxed()
+    }
+  }
+
+  // Ripped from lib.rs's test module
+  // TODO: Work out how to share things across tests
+  fn make_file(path: &Path, contents: &[u8], mode: u32) {
+    let mut file = std::fs::File::create(&path).unwrap();
+    file.write(contents).unwrap();
+    let mut permissions = std::fs::metadata(path).unwrap().permissions();
+    permissions.set_mode(mode);
+    file.set_permissions(permissions).unwrap();
+  }
+
+  // TODO: Is there a better way to do this?
```

Comment:
I couldn't find an obvious "Down-cast to known enum type" feature, but it could be that I'm just not using the right terminology... Any pointers would be appreciated :)

Reply:
I don't see a way either. https://stackoverflow.com/a/45473824

But it looks like in this case you don't really need to verify the type of the PathStat, since you just created the file/dir. So a bigger boilerplate-remover in here would be to add a `fn stat(p: &Path) -> Vec<PathStat>` test helper to replace both cast functions. I think you could do the equivalent of `VFS::expand(PathGlobs::create([p.to_string], [])).wait()`... maybe overkill, but not likely to slow the tests down a noticeable amount.